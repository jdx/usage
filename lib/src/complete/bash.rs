use heck::ToShoutySnakeCase;

pub fn complete_bash(bin: &str, usage_cmd: &str) -> String {
    // let usage = env::USAGE_BIN.display();
    let bin_up = bin.to_shouty_snake_case();
    // let bin = &spec.bin;
    // let raw = shell_escape::unix::escape(spec.to_string().into());
    format!(
        r#"
_{bin}() {{
    if ! command -v usage &> /dev/null; then
        echo >&2
        echo "Error: usage CLI not found. This is required for completions to work in {bin}." >&2
        echo "See https://usage.jdx.dev for more information." >&2
        return 1
    fi

    if [[ -z ${{_USAGE_SPEC_{bin_up}:-}} ]]; then
        _USAGE_SPEC_{bin_up}="$({usage_cmd})"
    fi

    COMPREPLY=( $(usage complete-word --shell bash -s "${{_USAGE_SPEC_{bin_up}}}" --cword="$COMP_CWORD" -- "${{COMP_WORDS[@]}}" ) )
    if [[ $? -ne 0 ]]; then
        unset COMPREPLY
    fi
    return 0
}}

shopt -u hostcomplete && complete -o nospace -o bashdefault -o nosort -F _{bin} {bin}
# vim: noet ci pi sts=0 sw=4 ts=4 ft=sh
"#
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[test]
    fn test_complete_bash() {
        assert_snapshot!(complete_bash("mycli", "mycli complete --usage").trim(), @r###"
        _mycli() {
            if ! command -v usage &> /dev/null; then
                echo >&2
                echo "Error: usage CLI not found. This is required for completions to work in mycli." >&2
                echo "See https://usage.jdx.dev for more information." >&2
                return 1
            fi

            if [[ -z ${_USAGE_SPEC_MYCLI:-} ]]; then
                _USAGE_SPEC_MYCLI="$(mycli complete --usage)"
            fi

            COMPREPLY=( $(usage complete-word --shell bash -s "${_USAGE_SPEC_MYCLI}" --cword="$COMP_CWORD" -- "${COMP_WORDS[@]}" ) )
            if [[ $? -ne 0 ]]; then
                unset COMPREPLY
            fi
            return 0
        }

        shopt -u hostcomplete && complete -o nospace -o bashdefault -o nosort -F _mycli mycli
        # vim: noet ci pi sts=0 sw=4 ts=4 ft=sh
        "###);
    }
}
