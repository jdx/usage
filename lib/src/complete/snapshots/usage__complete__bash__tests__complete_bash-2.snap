---
source: lib/src/complete/bash.rs
expression: "complete_bash(&CompleteOptions\n{\n    shell: \"bash\".to_string(), bin: \"mycli\".to_string(), cache_key:\n    Some(\"1.2.3\".to_string()), spec: None, usage_cmd:\n    Some(\"mycli complete --usage\".to_string()),\n})"
snapshot_kind: text
---
_mycli() {
    if ! command -v usage &> /dev/null; then
        echo >&2
        echo "Error: usage CLI not found. This is required for completions to work in mycli." >&2
        echo "See https://usage.jdx.dev for more information." >&2
        return 1
    fi

    if [[ -z ${_usage_spec_mycli_1_2_3:-} ]]; then
        _usage_spec_mycli_1_2_3="$(mycli complete --usage)"
    fi

    COMPREPLY=( $(usage complete-word --shell bash -s "${_usage_spec_mycli_1_2_3}" --cword="$COMP_CWORD" -- "${COMP_WORDS[@]}" ) )
    if [[ $? -ne 0 ]]; then
        unset COMPREPLY
    fi
    return 0
}

shopt -u hostcomplete && complete -o nospace -o bashdefault -o nosort -F _mycli mycli
# vim: noet ci pi sts=0 sw=4 ts=4 ft=sh
